### PR TITLE
`Development`: Add documentation for GitLab CI

### PR DESCRIPTION
**Pull request by bensofficial**
#### General
- [x] This is a small issue that I tested on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

### Description
This PR adds documentation on how to set up Artemis to work with GitLab and GitLab CI.

### Steps for Testing
Have a look at the generated documentation: https://artemis-platform--5807.org.readthedocs.build/en/5807/dev/setup/#gitlab-ci-and-gitlab-setup

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: docs/dev/setup/gitlabci-gitlab.rst.txt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        --url http://gitlab:80 \
+        --registration-token "PROJECT_REGISTRATION_TOKEN" \
+        --description "docker-runner" \
+        --maintenance-note "Test Runner" \
+        --tag-list "docker,artemis" \
+        --run-untagged="true" \
+        --locked="false" \
+        --access-level="not_protected"
+
+    **Note:** For local development, you might add ``--clone-url http://gateway.docker.internal:80`` or ``--clone-url http://172.17.0.1:80`` to the command above.
```
Comment (b-fein) [inline docs/dev/setup/gitlabci-gitlab.rst.txt:67]:
```suggestion
    .. note::
        For local development, you might add ``--clone-url http://gateway.docker.internal:80`` or ``--clone-url http://172.17.0.1:80`` to the command above.
```
This renders the note in a coloured block.

File: docs/dev/setup/gitlabci-gitlab.rst.txt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    You should now find the runner in the list of runners (``http://gitlab/admin/runners``)
+
```
Comment (b-fein) [inline docs/dev/setup/gitlabci-gitlab.rst.txt:70]:
```suggestion

.. note::
    Adding a runner in a production setup works the same way.
    The GitLab administration page also contains alternative ways of setting up GitLab runners.
    All variants should allow the passing of the configuration options ``tag-list``, ``run-untagged``, ``locked``, and ``access-level`` similarly as in the Docker command above.
    If forgotten, Artemis will not use this runner to run the tests for exercise submissions.


```

File: docs/dev/setup/gitlabci-gitlab.rst.txt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+Artemis
+^^^^^^^
+
+1. Generate authentication token
+    The notification plugin has to authenticate to upload the test results.
+    Therefore, a random token has to be generated.
```
Comment (b-fein) [inline docs/dev/setup/gitlabci-gitlab.rst.txt:76]:
```suggestion
    Therefore, a random string has to be generated, e.g., via a password generator.
    This should be used in place of ``notification-plugin-token`` value in the example config below.
```

File: docs/dev/setup/gitlabci-gitlab.rst.txt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+2. Configure Artemis
+    For local development, copy the following configuration into the ``application-local.yml`` file and adapt it with the values from the previous steps.
+    Please make sure, that the GitLab Runner can access Artemis via the URL specified.
```
Comment (b-fein) [inline docs/dev/setup/gitlabci-gitlab.rst.txt:79]:
```suggestion
    Please make sure that the GitLab Runner can access Artemis via the URL specified under ``artemis.version-control.url``.
```

**Review by b-fein — COMMENTED** [review]
Looks good. I’ve got a few suggestions where I think the documentation should be extended with some additional information.

File: docs/dev/setup/gitlabci-gitlab.rst.txt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        docker-compose -f src/main/docker/gitlab-gitlabci.yml --env-file src/main/docker/.env up --build -d
+
+3. Get your GitLab root password
+    .. code:: bash
+
+        docker exec -it gitlab grep 'Password:' /etc/gitlab/initial_root_password
+
+4. Generate an access token
+    Go to ``http://gitlab/-/profile/personal_access_tokens`` and generate an access token with all scopes.
+    This token is used in the Artemis configuration as `artemis.version-control.token`.
```
Comment (b-fein) [inline docs/dev/setup/gitlabci-gitlab.rst.txt:46]:
```suggestion
    This token is used in the Artemis configuration as ``artemis.version-control.token``.
```
italics → monospace

**Review by b-fein — APPROVED** [review]
Thank you! One small formatting suggestion, looks good otherwise.

**Review by b-fein — APPROVED** [review]

File: docs/dev/setup/gitlabci-gitlab.rst [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            user-management:
+                use-external: false
+                internal-admin:
+                    username: artemis_admin
+                    password: artemis_admin
+                accept-terms: false
+                login:
+                    account-name: TUM
+            version-control:
+                url: http://gitlab # TODO
```
Comment (Strohgelaender) [inline docs/dev/setup/gitlabci-gitlab.rst:113]: Instead of `TODO` maybe write e.g. `change this value`.

**Review by b-fein — APPROVED** [review]

**Review by tunargs — APPROVED** [review]
Looks great

**Review by Mtze — CHANGES_REQUESTED** [review]
I followed your tutorial to setup a local test environment. Unfortunately it doesn't work for me yet. Even with the "fixes" in the inline comments, I still can't execute a build since the spring config requires the gitlab url to be `localhost` to create the repository in the gitlab container but has to be `gitlab` for the runner to be able to clone properly. (even the clone url workaround seems to not work for me) 

Do you have a working local setup?

File: docs/dev/setup/gitlabci-gitlab.rst [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            user-management:
+                use-external: false
+                internal-admin:
+                    username: artemis_admin
+                    password: artemis_admin
+                accept-terms: false
+                login:
+                    account-name: TUM
+            version-control:
+                url: http://gitlab # change this value
```
[the remark this answers is not in the document]
Comment (bensofficial) [reply docs/dev/setup/gitlabci-gitlab.rst]: Resolved in testing session. c.f. changes in commit 3682457

File: docs/dev/setup/gitlabci-gitlab.rst [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                token: gitlab-personal-access-token # change this value
+            continuous-integration:
+                build-timeout: 30
+                artemis-authentication-token-value: notification-plugin-token # change this value
+            git:
+                name: Artemis
+                email: artemis.in@tum.de
+        server:
+            port: 8080
+            url: http://artemis:8080 # change this value
```
[the remark this answers is not in the document]
Comment (bensofficial) [reply docs/dev/setup/gitlabci-gitlab.rst]: Resolved in testing session. c.f. changes in commit 3682457

File: docs/dev/setup/gitlabci-gitlab.rst [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+3. Get your GitLab root password
+    .. code:: bash
+
+        docker exec -it gitlab grep 'Password:' /etc/gitlab/initial_root_password
+
+4. Generate an access token
+    Go to ``http://gitlab/-/profile/personal_access_tokens`` and generate an access token with all scopes.
+    This token is used in the Artemis configuration as ``artemis.version-control.token``.
+
```
[the remark this answers is not in the document]
Comment (bensofficial) [reply docs/dev/setup/gitlabci-gitlab.rst:52]: Resolved in testing session. c.f. changes in commit 3682457

**Pull request merged**